src/core/mesh/qgsmeshtracerenderer.cpp

```diff
@@ -0,0 +1,605 @@
+                            mDatasetValues.value( face[1] ).y() );
+
+  QgsVector v3 = QgsVector( mDatasetValues.value( face[2] ).x(),
+                            mDatasetValues.value( face[2] ).y() );
+
+
+  return QgsMeshLayerUtils::interpolateVectorFromVerticesData(
+           p1,
+           p2,
+           p3,
+           v1,
+           v2,
+           v3,
+           point );
+}
+
+QgsMeshVectorValueInterpolator::QgsMeshVectorValueInterpolator( QgsTriangularMesh triangularMesh,
+    QgsMeshDataBlock datasetVectorValues,
+    QgsMeshDataBlock scalarActiveFaceFlagValues ):
+  mTriangularMesh( triangularMesh ),
+  mDatasetValues( datasetVectorValues ),
+  mActiveFaceFlagValue( scalarActiveFaceFlagValues )
+{
+
+}
+
+QgsMeshVectorValueInterpolator::~QgsMeshVectorValueInterpolator() {}
+
+void QgsMeshVectorValueInterpolator::updateFace( const QgsPointXY &point )
+{
+  if ( ! QgsMeshUtils::isInTriangleFace( point, face, mTriangularMesh.vertices() ) )
+  {
+    /*the first version of this method uses QgsGeometry::contain to check if the point is in triangles,
+    *  not too time consuming for a simple triangle?
+    * why not using simpler function for this check ?  version 2 is  nearly 10 time faster.
+    */
+    mCacheFaceIndex = mTriangularMesh.faceIndexForPoint_v2( point );
+  }
+
+}
+
+QSize QgsMeshTraceField::size() const
+{
+  return mFieldSize;
+}
+
+QPoint QgsMeshTraceField::topLeft() const
+{
+  return mFieldTopLeftInDeviceCoordinates;
+}
+
+int QgsMeshTraceField::particleWidth() const
+{
+  return mFieldResolution;
+}
+
+QgsPointXY QgsMeshTraceField::positionToMapCoordinnate( const QPoint &pixelPosition, const QgsPointXY &positionInPixel )
+{
+  QgsPointXY mapPoint = mMapToFieldPixel.toMapCoordinates( pixelPosition );
+  mapPoint = mapPoint + QgsVector( positionInPixel.x() * mMapToFieldPixel.mapUnitsPerPixel(),
+                                   positionInPixel.y() * mMapToFieldPixel.mapUnitsPerPixel() );
+  return mapPoint;
+}
+
+QgsMeshTraceField::QgsMeshTraceField( const QgsRenderContext &renderContext, QgsMeshVectorValueInterpolator *interpolator, const QgsRectangle &layerExtent, double Vmax ):
+  mLayerExtent( layerExtent ),
+  mRenderContext( renderContext ),
+  mVectorValueInterpolator( interpolator ),
+  mVmax( Vmax )
+{
+}
+
+void QgsMeshTraceField::updateSize()
+{
+  const QgsMapToPixel &deviceMapToPixel = mRenderContext.mapToPixel();
+
+  QgsRectangle layerExtent = mRenderContext.coordinateTransform().transform( mLayerExtent );
+  QgsRectangle mapExtent = mRenderContext.extent();
+
+  QgsRectangle interestZoneExtent = layerExtent.intersect( mapExtent );
+  if ( interestZoneExtent == QgsRectangle() )
+  {
+    mValid = false;
+    mFieldSize = QSize();
+    mFieldTopLeftInDeviceCoordinates = QPoint();
+    initField();
+    return;
+  }
+
+  QgsPointXY interestZoneTopLeft = deviceMapToPixel.transform( QgsPointXY( interestZoneExtent.xMinimum(), interestZoneExtent.yMaximum() ) );
```

Comment:
when transforming points you probably need to `catch ( QgsCsException &cse )` in case the tranformation is invalid/cannot be done.

Reply:
yes

Reply:
is this implemented?